packages/flutter/lib/src/material/scaffold.dart

```diff
@@ -36,18 +36,63 @@ enum _ScaffoldSlot {
   statusBar,
 }
 
+/// Geometry information for scaffold components.
+///
+/// To get a [ValueNotifier] for the scaffold geometry call
+/// [Scaffold.geometryOf].
+@immutable
+class ScaffoldGeometry {
+  const ScaffoldGeometry({
+    this.bottomNavigationBarTop,
+    this.floatingActionButton,
```

Comment:
I think we're going to need two different types of ScaffoldGeometry.  The fab placement logic I'm working on depends on the scaffold's geometry for figuring out what offset to give the fab, so the position rect won't apply.

I think the easiest thing to do will be to pass the fab positioner an incomplete scaffold geometry, then we can finish populating it with the fab layout after it's computed.  

### TLDR
So long as it's possible to build ScaffoldGeometry without data about the fab's layout and without passing the incomplete object into the value notifier, this shouldn't be a problem.

Reply:
Will we be able to populate the floatingActionButton with it's position before paint time?

As long as by paint time we populated the fab information in the geometry object it should be ok.
In current CL we already populate it with incomplete data from the FabTransition before the first layout.

Reply:
Yes.  The way it works in #14368 currently is:

1. Scaffold passes FabPositioner and FabMotionAnimator to the Layout delegate
2. Layout delegate computes layout of everything but the fab and creates a ScaffoldGeometry
3. FabPositioner uses the ScaffoldGeometry to determine the Fab's Offset
4. Layout delegate computes the Fab's layout based on the requested offset (interpolated by animation when necessary)

The output of this can be a ScaffoldGeometry that includes the Fab's layout rect, and that will be accessible as part of the Layout delegate's layout phase, before painting.

Reply:
great

Reply:
it seems weird to me to use the same struct for computing the fab position and for obtaining the fab position